core/src/main/java/io/grpc/internal/ConnectivityStateManager.java

```diff
@@ -31,66 +31,89 @@
 
 package io.grpc.internal;
 
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
 import io.grpc.ConnectivityState;
+import io.grpc.ManagedChannel;
 import java.util.ArrayList;
 import java.util.concurrent.Executor;
+import javax.annotation.Nonnull;
 import javax.annotation.concurrent.NotThreadSafe;
 
 /**
- * Book-keeps the connectivity state and callbacks.
+ * Manages connectivity states of the channel. Used for {@link ManagedChannel#getState} to read the
+ * current state of the channel, for {@link ManagedChannel#notifyWhenStateChanged} to add
+ * listeners to state change events, and for {@link io.grpc.LoadBalancer.Helper#updateBalancingState
+ * LoadBalancer.Helper#updateBalancingState} to update the state and run the {@link #updateState}s.
  */
 @NotThreadSafe
-class ConnectivityStateManager {
-  private ArrayList<StateCallbackEntry> callbacks;
-
-  private ConnectivityState state;
+public final class ConnectivityStateManager {
+  private ArrayList<Listener> listeners = new ArrayList<Listener>();
 
-  ConnectivityStateManager(ConnectivityState initialState) {
-    state = initialState;
-  }
+  private volatile ConnectivityState state = ConnectivityState.IDLE;
 
-  void notifyWhenStateChanged(Runnable callback, Executor executor, ConnectivityState source) {
-    StateCallbackEntry callbackEntry = new StateCallbackEntry(callback, executor);
+  /**
+   * Adds a listener for state change event.
+   *
+   * <p>Must be synchronized with {@link #updateState}.
+   *
+   * <p>The {@code executor} must be one that can run RPC call listeners.
+   */
+  public void addListener(Runnable callback, Executor executor, ConnectivityState source) {
```

Comment:
I think `notifyWhenStateChanged` is more accurate, as it implies the listener is one-shot. `addLIstener` sounds like the listener will stay after being called.